server/athenian/api/internal/features/entries.py

```diff
@@ -376,7 +377,12 @@ async def calc_pull_request_metrics_line_github(
         )
         lines_grouper = partial(group_prs_by_lines, lines)
         repo_grouper = partial(group_by_repo, PullRequest.repository_full_name.name, repositories)
-        with_grouper = partial(group_prs_by_participants, participants)
+        # as optimization, if len(participants) == 1, we've already filtered in SQL
+        # so don't have to re-check
+        if len(participants) == 1:
```

Comment:
imho it's simpler to have the optimization outside, else group_prs_by_participants contract is more complex

Reply:
Let's move it inside. I was there, and it's hell once the number of these optimizations grow. It is trivial to move `if` inside `group_prs_by_participants` and, honestly, it's their business to work correctly. Besides, we have similar `if`-s in other groupers.

Reply:
If we keep it  inside I need to add a boolean flag `skip_single_group_optimization` to `group_prs_by_participants`, is this better ?

Reply:
This is because in batch_calc_prs... to SQL we passed multiple participant groups, but then for a single request we group by a single participant group